File: client/src/main/java/com/fiwio/iot/demeter/app/demeter/di/ApplicationComponent.java
package com.fiwio.iot.demeter.app.demeter.di;

import com.fiwio.iot.demeter.app.demeter.addtask.di.AddSchedulerTaskComponent;
import com.fiwio.iot.demeter.app.demeter.addtask.di.AddSchedulerTaskModule;
import com.fiwio.iot.demeter.app.demeter.main.MainActivity;
import com.fiwio.iot.demeter.app.demeter.remote.di.RemoteControlComponent;
import com.fiwio.iot.demeter.app.demeter.remote.di.RemoteControlModule;
import com.fiwio.iot.demeter.app.demeter.scheduler.di.SchedulerComponent;
import com.fiwio.iot.demeter.app.demeter.scheduler.di.SchedulerModule;
import com.fiwio.iot.demeter.app.demeter.splash.di.SplashComponent;
import com.fiwio.iot.demeter.app.demeter.splash.di.SplashModule;
import com.fiwio.iot.demeter.app.di.NetModule;

import javax.inject.Singleton;

import dagger.Component;

@Singleton
@Component(
        modules = {ApplicationModule.class, NetModule.class}
)
public interface ApplicationComponent {
    RemoteControlComponent plus(RemoteControlModule remoteControlModule);

    SchedulerComponent plus(SchedulerModule schedulerModule);

    SplashComponent plus(SplashModule remoteControlModule);

    AddSchedulerTaskComponent plus(AddSchedulerTaskModule addSchedulerTaskModule);

    void inject(MainActivity mainActivity);
}

File: mobile-ui/src/main/java/com/fiwio/iot/demeter/android/ui/injection/ApplicationComponent.kt
package com.fiwio.iot.demeter.android.ui.injection

import android.app.Application
import com.fiwio.iot.demeter.android.ui.app.DemeterApplication
import com.fiwio.iot.demeter.android.ui.feature.firebase.di.FirebaseComponent
import com.fiwio.iot.demeter.android.ui.feature.firebase.di.FirebaseModule
import com.fiwio.iot.demeter.android.ui.feature.main.di.MainComponent
import com.fiwio.iot.demeter.android.ui.feature.main.di.MainModule
import com.fiwio.iot.demeter.android.ui.feature.main.di.OfflineModule
import com.fiwio.iot.demeter.android.ui.feature.offline.di.OfflineComponent
import com.fiwio.iot.demeter.android.ui.feature.splash.di.SplashComponent
import com.fiwio.iot.demeter.android.ui.feature.splash.di.SplashModule
import dagger.BindsInstance
import dagger.Component
import javax.inject.Singleton

@Singleton
@Component(modules = arrayOf(ApplicationModule::class, DataModule::class))
interface ApplicationComponent {

    fun inject(demeterApplication: DemeterApplication)

    fun plus(mainModule: MainModule): MainComponent
    fun plus(offlineModule: OfflineModule): OfflineComponent
    fun plus(splashModule: SplashModule): SplashComponent
    fun plus(firebaseModule: FirebaseModule): FirebaseComponent


    @Component.Builder
    interface Builder {
        @BindsInstance
        fun application(application: Application): Builder

        fun build(): ApplicationComponent
    }
}
